### ISSUE TITLE
NetDevice Attributes: Overhaul

### ISSUE THREAD

**Issue by coxley** [trigger/trigger#232]
### Description

NetDevices needs a pretty good overhaul as far as tags/attributes are concerned. We have talked about this a handful of times, but I want to get it on "paper".
### Context

Here's a list of some current issues:
1. Baked in attributes
   - There are specific attributes that Trigger expects/requires you to have. There are static properties in the NetDevices object. This ties into problem no. 2
2. NetDevices supports arbritrary attributes and can fetch via nd.search!...kind of
   - If you add arbritrary attributes to your NetDevice loader (JSON, SQLite, etc) then **every** device must have it and not be null or else it will non exist on any object. (Specifically, the first device that gets loaded from the loader on NetDevices() instantiation sets the standard for the rest, which last time I tested 1yr ago it wasn't immediately predictable)
3. Attributes utilize `CamelCase` naming because javascript/xml/legacy.
   - `lowcase_underscore` attributes are generally preferred in the Python community. We've talked about changing it to be lowercase and providing a `MAP` for backward compatibility, though I have a better idea that won't involve butting into peoples lives (ok maybe less lives, more way of working)
### Proposal

Everyone wants to do things differently for different reasons. I want Trigger to have `underscore_separated` attributes because I'm a Python fan in addition to a NSoT fan for storing device resources. May not be the case for other people and their existing circumstances.

I propose instead we do something similar to what we do for Commando plugins and NetDevice loaders. Let's define our required attributes (manufacturer, platform, node name, so on) and let the user specify what these attribute "roles" correlate to.

Caring about the actual attribute name a user wants to use seems pointless in the long run now that I think about it and changing this will fix all the problems above, expanding the modularity of Trigger. We should also be able to access all the attributes from the NetDevice object via properties, just enabling protection for conflicting with method names.

What's the consensus for this?

**Comment by jathanism** [comment]
We need to define the minimum set of fields. I think currently it's really only:

* manufacturer (aka vendor) - It should really just be `vendor`
* deviceType - It could literally just be `type`
* nodeName - It should be `hostname`
* adminStatus - This is up/down status, perhaps just `admin_status`

We could probably literally do away with every other 'pre-defined' field, however, we need to take them all into consideration. For example, with the chatter around OOB support, do we want to maintain `OOBTerminalServerFQDN` and `OOBTerminalServerTCPPort` (and if so obviously they should also be renamed)? Should they be optional?

Lastly, we need to make sure we're considering migration from older versions and/or backwards-compatibility, by way of a configurable mapping from "legacy" to "current" fields. This would also allow adaptability from environments where the field name mappings differ and prevent newcomers from having to munge their incoming source data to something Trigger expects, by instead just being able to specify it in their configuration.

**Comment by coxley** [comment]
Agreed about compatibility. Would be nice if it was done in a way easy to phase out later vs having legacy shit in the codebase forever. 

I think this should be on the roadmap for 1.6 because I like to be optimistic :)

I'm a fan of your renames but would still like that to be configurable context. Eg 

```python
ATTR_ROLES = {
    'vendor': 'who_to_blame',
    'admin_status': 'status',
}

```

OOB role should be optional yeah, but looked for.

**Comment by jathanism** [comment]
**v2.0.0 Update:** The CamelCase attribute naming from the XML/JavaScript legacy is still present in v2.0.0. The NetDevice attribute system has not been overhauled yet.

The proposal to decouple attribute "roles" from their names (similar to the plugin architecture used for Commando and NetDevice loaders) remains valid and would be a significant improvement. This would allow users to define their own attribute mappings rather than being locked into the current schema.

Keeping this open as a longer-term enhancement goal.
